core/src/main/java/dev/streamx/cli/ingestion/BaseCommand.java

```diff
@@ -0,0 +1,74 @@
+package dev.streamx.cli.ingestion;
+
+import dev.streamx.cli.exception.IngestionClientException;
+import dev.streamx.clients.ingestion.StreamxClient;
+import dev.streamx.clients.ingestion.exceptions.StreamxClientException;
+import jakarta.inject.Inject;
+import java.util.List;
+import org.apache.commons.lang3.StringUtils;
+import picocli.CommandLine.ArgGroup;
+import picocli.CommandLine.Model.CommandSpec;
+import picocli.CommandLine.Spec;
+
+public abstract class BaseCommand implements Runnable {
+
+  @ArgGroup(exclusive = false)
+  protected IngestionArguments ingestionArguments;
+
+  @Spec
+  protected CommandSpec spec;
+
+  @Inject
+  protected SchemaProvider schemaProvider;
+
+  @Inject
+  protected StreamxClientProvider streamxClientProvider;
+
+  protected abstract void performCommand(StreamxClient client) throws StreamxClientException;
+
+  @Override
+  public void run() {
+    StreamxClient client = createClient();
```

Comment:
@jgardo-ds , my comment was lost after renaming this class.

Now we cannot use try-with-resources, since when the publish operation fails - the client closes its underlying httpClient.

So we would not able be to call it to get the list of valid channels in the exception handler below.

That's why manual lifetime management of the client instance had to be introduced.

But there are plans for bigger refactorings in all streamx / cli / ingestion projects, as proposed by @michalcukierman - to receive clean implementation

Reply:
Yes, I saw your previous answer, so I deleted this comment.